RFD: Long-lived Machine ID local SSH proxy (#42546)
gravitational/teleport: rfd/0172-tbot-ssh-muxer.md

socket, the client will write a short message containing the target host and
port. The service will then establish the connection, and then begin forwarding
data between the socket and the upstream connection.

The client will be configured as the `ProxyCommand`, and additionally
`ProxyUseFdpass` will be enabled. The `ProxyUseFdPass` allows the `ProxyCommand`
to return an FD via STDOUT using `sendmsg`, rather than the `ProxyCommand`
actively forwarding the connection. This means the client can exit after the
connection to the local proxy is established.

This solution means that whilst a process will be spun up for each SSH
connection, it will return after the connection is established. This reduces the
long-term overhead associated with each connection and the impact on the Linux
scheduler.

#### Protocol

The Unix socket will implement a simple protocol:

1. Upon connection, the client will send a message indicating the target.
  a. The message will be encoded in JSON. This will allow more fields to be 
    added in the future without breaking compatability.
  b. The `host_port` field will be a string in the format `host:port`.

> 2. The server will establish a connection to the target, and then begin

Author: @espadolini do we need to have some way of indicating failure messages ? Or are we happy to just close the connection and log on the serverside.

   forwarding data between the local connection and the target.

Also considered was implementing the SOCKS5 protocol. Whilst this is more 
standardized, it is more complicated than necessary. It also relies on our
existing implementation of a SOCKS5 server which has not received signficant
attention in some time.

#### Client Binary

As it stands today, the `tbot` binary has significant overhead associated with
it.

Due to this, we will introduce a new, extremely lightweight binary that will

> act as the client.

Author: I discussed this with Edoardo - since there's no reason we could have a subcommand by default and then users can switch to the more performant standalone if they want. It'll definitely put a dink in the performance but I'm unsure if @espadolini has done a measurement on that.

The client will accept two arguments:

1. The path to the Unix socket.
2. The message to send to the `tbot` proxy service. This will have the host and
   port templated into it by OpenSSH with the `%h` and `%p` placeholders.

Comparing Rust and Go for this lightweight binary:

- Rust has a smaller binary size: 400KiB vs 1.8MiB.

> - Rust executes faster.

Reviewer: The actual code is still like 4 or 5 syscalls including `exit()` in both cases so "speed of execution" is not really different I don't think, so I'd say something like 
```suggestion
- Rust has little to no overhead, whilst Go requires spinning up a runtime with multiple OS threads and a garbage collector before any user code gets to run.
```

- Tuning Go's GC behaviour requires also using `env` as part of the

>   ProxyCommand. This is not necessary with Rust.

Reviewer: We could set parameters during some `init()` too, but I'm not sure if they'd cover everything and if it changes our decision, ultimately.

Author: Yeah - I'm not sure it makes a major difference either way but I think the very fact that we want to tune the GC is not a good sign for Go.

For this binary, we will use Rust.

### Initial Scope

For the initial version, we must include:

- The long-lived proxying background service within `tbot`
- The new lightweight client binary with support for ProxyUseFdPass.
- A configuration option to control the maximum number of proxied connections.
  - When this maximum is reached, the service will wait until the number of

>     active connections drops below the maximum.

Reviewer: How is this in any way practical, with `ControlMaster`/`ControlPath`? Connections won't get freed until minutes later, and at that point you've basically just rejected the connection.

Author: Do you feel that we would be better off removing any form of limitation on the client side ? Or should we be more aggressive and reject the connections rather than holding off ?

Reviewer: Hard to say. Rejecting connections would be better, but blocking will let a bunch of short-lived connection through while respecting the concurrency limit, and both options can be good depending on the workload.

Do we need a connection limit?

Author: Ended up dropping the limit from implementation - I'll remove from RFD.

Author: [9866425](https://github.com/gravitational/teleport/pull/42546/commits/98664250527be2ab02dd8b914e25d4ab9486b195)

  - This will allow a sane limit to be provided to prevent the long-lived `tbot`
    process from being overwhelmed.
- Prometheus metrics which exposes the health of the proxying.
  - Number of active proxied connections.
  - Number of attempts to establish proxied connections, the time this has
    taken, and the success status.
- OpenTelemetry trace spans at key points.
  - This will allow us to target future performance improvements.

### Future Scope

#### Caching

We can further improve performance by caching the results of operations such as
resolving the target host from labels.

This work is deferred as it is complex and introduces risks. Information from
OpenTelemetry tracing will allow us to target the most impactful areas at a 
later date.

#### Upstream Connection Pooling

The initial version will not share the upstream connections to the Teleport

> Proxy.

Reviewer: Huh... why?

Author: I worry a little bit about performance issues introduced by head of line blocking etc or overuse of the shared connection. If we want to do this, we probably need to carve out some time to determine a safe number of downstream connections to upstream connections. If you're confident we can determine a safe value for this immediately then I'm happy to include this in the work.

Author: > Without connection pooling are we really going to reduce resource consumption?

Even without connection pooling we:

- Significantly reduce the work for the Linux scheduler by reducing the number of tasks/processes that are active
- Reducing the amount of memory consumed by the overhead of each of those tbots. Even with the improvements we've made to the heap size, we're probably still using a significant amount more memory than actually necessary to run run a single connection.

@espadolini  is going to run a test with one upstream connection per downstream connection - that'll give us a comparison point and we can make a firmer decision on whether or not the initial iteration needs to include connection sharing.

By "future improvement", I'm not really envisioning this in the far future - more just trying to tighten the scope on the initial version so we don't bite off more than we can chew. Effective connection pooling/sharing seems non-trivial to me - and if there's a significant improvement without adding it, then it seems to me to make more sense for us to work iteratively to keep complexity low. If we can deliver a significant improvement sooner and without complexity, it reduces the pressure on us and means we can spend the time we need ensure that the connection pooling/sharing is stable.

Author: Having gone through the figures with Edoardo - I think this is a big enough improvement that we can proceed straight to including a connection cycler.

Author: Addressed in [e141c4b](https://github.com/gravitational/teleport/pull/42546/commits/e141c4b06d710f67402f3d49078b7c069643599f)

Sharing upstream connections will allow us to reduce the number of connections
and reduce the latency and overhead associated with establishing a new
connection.

This work is deferred as it is complex and introduces risks. We should first
ensure that the initial version is stable and to take measurements which can

> inform the design of the connection pooling.

Author: 👍  I'll add this as one consideration.

### UX

The user will configure `ssh-local-proxy` in the `tbot` configuration file:

```yaml
services:
  - type: ssh-local-proxy
    # Destination is where the ssh_config and UDS will be written to. It must be
    # a directory.
    destination:
      type: directory
      path: /opt/machine-id
    # The maximum number of connections that can be proxied at once.
    max_connections: 100
```

### Security

#### Unix Domain Socket is unauthenticated

The Unix Domain Socket will be unauthenticated, and this means that any process
which is able to connect to it will be able to establish upstream connections.

Fortunately, existing Linux permissions mechanisms can be used to restrict 
access to the UDS and this is similar to how we handle the credentials
output by `tbot` today.

> We should ensure we document the best practices to secure the UDS.

Reviewer: Permissions on unix sockets are always kind of weird and there's no good way to set the permission of a socket as you're creating it except by playing with the umask (which is process-wide, yuck), so we should take care to use a _subdirectory_ with restrictive permissions that contains the socket - at which point the socket permissions don't matter beyond being at least `u+rw`.

Author: At least for `tbot`, we typically operate on the assumption that the user the bot runs as will often not be the same user that is consuming the credentials output by `tbot`. We handled this for the Workload Identity API by chmodding the newly created socket ( which obviously is a little racy) to 777 and then encouraging control by placing more restrictive permissions on the directory which is not recreated on each start.

Reviewer: The listening socket should still be in a dedicated subdirectory that we chmod 700 before doing anything else.

Reviewer: I'm fine with using `tbot`'s directory ACL management

We will introduce a long-lived service in `tbot`: `ssh-local-proxy`.

Upon start, the service will write a SSH config to a configured destination.
This SSH config should be used by OpenSSH to leverage the service.

Once initialized, the service will listen on a Unix socket. Upon connection to this
socket, the client will write a short message containing the target host and
port. The service will then establish the connection, and then begin forwarding
data between the socket and the upstream connection.

The client will be configured as the `ProxyCommand`, and additionally
`ProxyUseFdpass` will be enabled. The `ProxyUseFdPass` allows the `ProxyCommand`
to return an FD via STDOUT using `sendmsg`, rather than the `ProxyCommand`
actively forwarding the connection. This means the client can exit after the
connection to the local proxy is established.

This solution means that whilst a process will be spun up for each SSH
connection, it will return after the connection is established. This reduces the
long-term overhead associated with each connection and the impact on the Linux
scheduler.

#### Protocol

A simple protocol will exist for interacting with the multiplexer:

> 1. The server will open a Unix domain socket named `tbot_ssh_multiplexer.v1.sock`.

Reviewer: Characters are sort of at a premium when dealing with unix domain sockets; considering that we're in a tbot output directory for the ssh-multiplexer service already, why not `mux/v1.sock`?

Author: Sounds good to me

Author: a082be328b

2. Upon connection, the client will send a request indicating the target.
  a. The request will be encoded in JSON and be followed by a NUL (ASCII/Unicode 0) character.
  b. This JSON message will contain two fields:
    i. `host`: The target host.

>     ii. `port`: The target host's port.

Reviewer: What are we gaining with the JSON?

Author: I'm not extremely married to this decision - I went with JSON as it would make it trivial to add additional optional fields in the future as necessary without the need for breaking the API version. If there's some perceived risk with JSON, I'm more than happy to switch it back to just `host:port` terminated with NULL.

Once initialized, the service will listen on a Unix socket. Upon connection to this
socket, the client will write a short message containing the target host and
port. The service will then establish the connection, and then begin forwarding
data between the socket and the upstream connection.

The client will be configured as the `ProxyCommand`, and additionally
`ProxyUseFdpass` will be enabled. The `ProxyUseFdPass` allows the `ProxyCommand`
to return an FD via STDOUT using `sendmsg`, rather than the `ProxyCommand`
actively forwarding the connection. This means the client can exit after the
connection to the local proxy is established.

This solution means that whilst a process will be spun up for each SSH
connection, it will return after the connection is established. This reduces the
long-term overhead associated with each connection and the impact on the Linux
scheduler.

#### Protocol

The Unix socket will implement a simple protocol:

1. Upon connection, the client will send a message indicating the target.
  a. The message will be encoded in JSON. This will allow more fields to be 
    added in the future without breaking compatability.

>   b. The `host_port` field will be a string in the format `host:port`.

Author: Yah - I mean really this is just going to be whatever %h and %p is - so proxy templating will occur within the daemon to resolve that. I'll add anote.

This SSH config should be used by OpenSSH to leverage the service.

Once initialized, the service will listen on a Unix socket. Upon connection to this
socket, the client will write a short message containing the target host and
port. The service will then establish the connection, and then begin forwarding
data between the socket and the upstream connection.

The client will be configured as the `ProxyCommand`, and additionally
`ProxyUseFdpass` will be enabled. The `ProxyUseFdPass` allows the `ProxyCommand`
to return an FD via STDOUT using `sendmsg`, rather than the `ProxyCommand`
actively forwarding the connection. This means the client can exit after the
connection to the local proxy is established.

This solution means that whilst a process will be spun up for each SSH
connection, it will return after the connection is established. This reduces the
long-term overhead associated with each connection and the impact on the Linux
scheduler.

#### Protocol

A simple protocol will exist for interacting with the multiplexer:

1. The server will open a Unix domain socket named `tbot_ssh_multiplexer.v1.sock`.
2. Upon connection, the client will send a request indicating the target.

>   a. The request will be encoded in JSON and be terminated with NULL.

Reviewer: ```suggestion
  a. The request will be encoded in JSON and be followed by a NUL (ASCII/Unicode 0) character.
```

  b. This JSON message will contain two fields:
    i. `host`: The target host.
    ii. `port`: The target host's port.
3. The server will establish a connection to the target, and then begin
   forwarding data between the local connection and the target.

Also considered was implementing the SOCKS5 protocol. Whilst this is more 
standardized, it is more complicated than necessary. In addition, it seems like
there's not many tools in the wild that can connect to a SOCKS server over a

> unix domain socket, so any advantage gained by the use of a standard protocol

Reviewer: ```suggestion
UNIX domain socket, so any advantage gained by the use of a standard protocol
```

This work is deferred as it is complex and introduces risks. Information from
OpenTelemetry tracing will allow us to target the most impactful areas at a 
later date.

#### Further improving re-use of connections

The initial version will use a rudimentary connection "cycler". This will cause
a new connection to be established after a certain number of downstream
connections have been established. It will then close the upstream connection
once all downstream connections using it have been closed.

This could be further improved by balancing incoming downstream connections
against a pool of upstream connections, because the upstream connections would
live longer and we'd avoid a scenario where an upstream connection remains open
but is only being utilised by one or two remaining downstream connections.

### UX

The user will configure `ssh-local-proxy` in the `tbot` configuration file:

```yaml
services:
  - type: ssh-local-proxy
    # Destination is where the ssh_config and UDS will be written to. It must be
    # a directory.
    destination:
      type: directory
      path: /opt/machine-id
    # The maximum number of connections that can be proxied at once.

>     max_connections: 100

Reviewer: Is this the limit per grpc connection, or global for the whole service?

Author: [9866425](https://github.com/gravitational/teleport/pull/42546/commits/98664250527be2ab02dd8b914e25d4ab9486b195)
